File: scripts/R/02_build_treatment.R
## ============================================================================
## 02_build_treatment.R
## Build master DC inventory: merge parcels + buildings, clean dates,
## compute centroids, assign census tracts
## ============================================================================

set.seed(20260329)
library(sf)
library(dplyr)
library(lubridate)

raw_dir  <- here::here("explorations", "data_collection", "raw")
out_dir  <- here::here("explorations", "data_collection", "processed")
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)

# --- 1. Load data ---
message("Loading DC parcels...")
parcels <- st_read(file.path(raw_dir, "dc_existing_parcels.geojson"), quiet = TRUE)
message("  ", nrow(parcels), " parcels loaded")

message("Loading DC buildings...")
buildings <- st_read(file.path(raw_dir, "dc_existing_permitted_buildings.geojson"), quiet = TRUE)
message("  ", nrow(buildings), " buildings loaded")

message("Loading census tracts...")
tracts <- st_read(file.path(raw_dir, "census_tracts_va", "tl_2020_51_tract.shp"), quiet = TRUE)
loudoun_tracts <- tracts %>% filter(COUNTYFP == "107")
message("  ", nrow(loudoun_tracts), " Loudoun County tracts")

# --- 2. Clean building permit dates ---
clean_permit_dates <- function(epoch_ms) {
  dates <- as.Date(as.POSIXct(epoch_ms / 1000, origin = "1970-01-01", tz = "UTC"))
  # Flag implausible dates (before 1990 or after 2030)
  dates[!is.na(dates) & (dates < as.Date("1990-01-01") | dates > as.Date("2030-01-01"))] <- NA
  return(dates)
}

buildings <- buildings %>%
  mutate(
    permit_date = clean_permit_dates(BP_ISSUE_DT),
    year_built_clean = ifelse(YEAR_BUILT > 1900 & YEAR_BUILT < 2030, YEAR_BUILT, NA_real_)
  )

message("\nBuilding date summary:")
message("  Valid permit dates: ", sum(!is.na(buildings$permit_date)), " / ", nrow(buildings))
message("  Date range: ", min(buildings$permit_date, na.rm = TRUE), " to ",
        max(buildings$permit_date, na.rm = TRUE))
message("  Valid year_built: ", sum(!is.na(buildings$year_built_clean)), " / ", nrow(buildings))
message("  YEAR_BUILT=1000 (placeholder): ", sum(buildings$YEAR_BUILT == 1000, na.rm = TRUE))

# --- 3. Aggregate buildings to parcel level ---
bldg_agg <- buildings %>%
  st_drop_geometry() %>%
  group_by(AM_MCPI) %>%
  summarize(
    n_buildings            = n(),
    total_sqft             = sum(as.numeric(SQUARE_FEET), na.rm = TRUE),
    earliest_permit_date   = min(permit_date, na.rm = TRUE),
    latest_permit_date     = max(permit_date, na.rm = TRUE),
    earliest_year_built    = min(year_built_clean, na.rm = TRUE),
    n_built                = sum(BUILT_STATUS == "BUILT", na.rm = TRUE),
    n_under_construction   = sum(BUILT_STATUS == "UNDER CONSTRUCTION", na.rm = TRUE),
    addresses              = paste(unique(na.omit(AM_BASE_ADDRESS)), collapse = "; "),
    .groups = "drop"
  ) %>%
  # Fix Inf/-Inf from empty groups
  mutate(
    across(where(is.numeric), ~ ifelse(is.infinite(.), NA_real_, .)),
    earliest_permit_date = if_else(is.infinite(as.numeric(earliest_permit_date)),
                                   as.Date(NA), earliest_permit_date)
  )

message("\nAggregated to ", nrow(bldg_agg), " unique MCPIs from buildings")

# --- 4. Compute parcel centroids ---
# Project to Virginia State Plane South (meters) for accurate centroids
parcels_proj <- st_transform(parcels, 32147)
centroids_proj <- st_centroid(parcels_proj)
centroids_4326 <- st_transform(centroids_proj, 4326)

coords <- st_coordinates(centroids_4326)
parcels$lon <- coords[, 1]
parcels$lat <- coords[, 2]

# --- 5. Join parcels to building aggregates ---
inventory <- parcels %>%
  st_drop_geometry() %>%
  left_join(bldg_agg, by = c("PA_MCPI" = "AM_MCPI"))

message("\nJoin results:")
message("  Parcels with building data: ", sum(!is.na(inventory$n_buildings)), " / ", nrow(inventory))
message("  Parcels without building data: ", sum(is.na(inventory$n_buildings)))

# --- 6. Assign census tracts ---
centroids_sf <- st_as_sf(inventory, coords = c("lon", "lat"), crs = 4326)
centroids_sf <- st_transform(centroids_sf, st_crs(loudoun_tracts))
tract_join <- st_join(centroids_sf, loudoun_tracts %>% select(TRACTCE, GEOID), join = st_within)

inventory$census_tract <- tract_join$TRACTCE
inventory$census_geoid <- tract_join$GEOID

message("  Parcels with census tract: ", sum(!is.na(inventory$census_tract)), " / ", nrow(inventory))

# --- 7. Select and rename output columns ---
dc_inventory <- inventory %>%
  transmute(
    mcpi             = PA_MCPI,
    project          = Project,
    owner            = Owner,
    ownership_cat    = Ownership_Category,
    lon, lat,
    overall_sqft     = as.numeric(Overall_SQ_FT),
    built_status     = Built_Status,
    zoning           = ZONING,
    place_type       = PLACE_TYPE,
    policy_area      = POLICY_AREA,
    gis_acres        = PA_GIS_ACRE,
    n_buildings,
    total_bldg_sqft  = total_sqft,
    earliest_permit  = earliest_permit_date,
    latest_permit    = latest_permit_date,
    earliest_year    = earliest_year_built,
    n_built,
    n_under_construction,
    addresses,
    census_tract,
    census_geoid
  )

# --- 8. Save ---
out_file <- file.path(out_dir, "dc_master_inventory.csv")
write.csv(dc_inventory, out_file, row.names = FALSE)

message("\n=== DC Master Inventory ===")
message("Rows: ", nrow(dc_inventory))
message("Saved to: ", out_file)
message("\nSummary by built status:")
print(table(dc_inventory$built_status, useNA = "ifany"))
message("\nPermit date range: ", min(dc_inventory$earliest_permit, na.rm = TRUE),
        " to ", max(dc_inventory$latest_permit, na.rm = TRUE))
message("Total buildings: ", sum(dc_inventory$n_buildings, na.rm = TRUE))
message("Total building sqft: ", format(sum(dc_inventory$total_bldg_sqft, na.rm = TRUE), big.mark = ","))
message("\nZoning breakdown:")
print(sort(table(dc_inventory$zoning), decreasing = TRUE))

File: scripts/python/02_build_treatment.py
"""
02_build_treatment.py
Build master DC inventory: merge parcels + buildings, clean dates,
compute centroids, assign census tracts.

Uses only Python stdlib (json, csv, math, datetime) — no geopandas/shapely needed.
GeoJSON polygons → centroids computed as simple coordinate averages (sufficient
for small polygons like parcels at this latitude).
"""

import json
import csv
import math
from datetime import datetime, timezone
from pathlib import Path
from collections import defaultdict

ROOT = Path(__file__).resolve().parents[2]
RAW = ROOT / "explorations" / "data_collection" / "raw"
OUT = ROOT / "explorations" / "data_collection" / "processed"
OUT.mkdir(parents=True, exist_ok=True)


def polygon_centroid(geometry):
    """Approximate centroid from GeoJSON geometry (mean of exterior ring coords)."""
    if geometry["type"] == "Polygon":
        ring = geometry["coordinates"][0]
    elif geometry["type"] == "MultiPolygon":
        # Use largest polygon by number of vertices
        ring = max(geometry["coordinates"], key=lambda p: len(p[0]))[0]
    else:
        return None, None
    lons = [c[0] for c in ring]
    lats = [c[1] for c in ring]
    return sum(lons) / len(lons), sum(lats) / len(lats)


def epoch_ms_to_date(epoch_ms):
    """Convert epoch milliseconds to ISO date string. Returns None for invalid."""
    if epoch_ms is None:
        return None
    try:
        dt = datetime.fromtimestamp(epoch_ms / 1000, tz=timezone.utc)
        if dt.year < 1990 or dt.year > 2030:
            return None
        return dt.strftime("%Y-%m-%d")
    except (ValueError, OSError, OverflowError):
        return None


def haversine_km(lon1, lat1, lon2, lat2):
    """Haversine distance in km between two WGS84 points."""
    R = 6371.0
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (math.sin(dlat / 2) ** 2 +
         math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) *
         math.sin(dlon / 2) ** 2)
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def point_in_polygon_simple(px, py, polygon_coords):
    """Ray-casting point-in-polygon test for a single ring."""
    n = len(polygon_coords)
    inside = False
    j = n - 1
    for i in range(n):
        xi, yi = polygon_coords[i]
        xj, yj = polygon_coords[j]
        if ((yi > py) != (yj > py)) and (px < (xj - xi) * (py - yi) / (yj - yi) + xi):
            inside = not inside
        j = i
    return inside


# --- 1. Load DC parcels ---
print("Loading DC parcels...")
with open(RAW / "dc_existing_parcels.geojson") as f:
    parcels_data = json.load(f)
parcels = parcels_data["features"]
print(f"  {len(parcels)} parcels loaded")

# --- 2. Load DC buildings ---
print("Loading DC buildings...")
with open(RAW / "dc_existing_permitted_buildings.geojson") as f:
    buildings_data = json.load(f)
buildings = buildings_data["features"]
print(f"  {len(buildings)} buildings loaded")

# --- 3. Clean building dates and aggregate to parcel level ---
bldg_by_mcpi = defaultdict(list)
valid_dates = 0
placeholder_years = 0

for b in buildings:
    p = b["properties"]
    mcpi = p.get("AM_MCPI")
    if not mcpi:
        continue

    permit_date = epoch_ms_to_date(p.get("BP_ISSUE_DT"))
    if permit_date:
        valid_dates += 1

    year_built = p.get("YEAR_BUILT")
    if year_built == 1000:
        placeholder_years += 1
        year_built = None
    elif year_built and (year_built < 1900 or year_built > 2030):
        year_built = None

    bldg_by_mcpi[mcpi].append({
        "permit_date": permit_date,
        "year_built": year_built,
        "sqft": p.get("SQUARE_FEET", 0) or 0,
        "built_status": p.get("BUILT_STATUS", ""),
        "address": (p.get("AM_BASE_ADDRESS") or "").strip(),
    })

print(f"\nBuilding date summary:")
print(f"  Valid permit dates: {valid_dates} / {len(buildings)}")
print(f"  YEAR_BUILT=1000 (placeholder): {placeholder_years}")
print(f"  Unique MCPIs with buildings: {len(bldg_by_mcpi)}")

# --- 4. Aggregate buildings per parcel ---
def aggregate_buildings(bldg_list):
    dates = [b["permit_date"] for b in bldg_list if b["permit_date"]]
    years = [b["year_built"] for b in bldg_list if b["year_built"]]
    addrs = list(set(b["address"] for b in bldg_list if b["address"]))
    return {
        "n_buildings": len(bldg_list),
        "total_bldg_sqft": sum(b["sqft"] for b in bldg_list),
        "earliest_permit": min(dates) if dates else None,
        "latest_permit": max(dates) if dates else None,
        "earliest_year": min(years) if years else None,
        "n_built": sum(1 for b in bldg_list if b["built_status"] == "BUILT"),
        "n_under_construction": sum(1 for b in bldg_list if b["built_status"] == "UNDER CONSTRUCTION"),
        "addresses": "; ".join(addrs[:3]),
    }

bldg_agg = {mcpi: aggregate_buildings(bl) for mcpi, bl in bldg_by_mcpi.items()}

# --- 5. Load census tracts (filter to Loudoun = COUNTYFP 107) ---
print("\nLoading census tracts...")
# Census tracts are in shapefile; read the GeoJSON we can create from the DBF
# Actually, we have the TIGER shapefile but can't read .shp without libraries.
# Instead, use a simple approach: we know Loudoun FIPS = 51107.
# For tract assignment, we'll note it as TBD and fill when geo libraries available.
tract_assignment_available = False

# --- 6. Build inventory ---
print("\nBuilding master inventory...")
inventory = []

for feat in parcels:
    p = feat["properties"]
    mcpi = p.get("PA_MCPI", "")
    lon, lat = polygon_centroid(feat["geometry"])

    bldg = bldg_agg.get(mcpi, {})

    row = {
        "mcpi": mcpi,
        "project": p.get("Project", ""),
        "owner": p.get("Owner", ""),
        "ownership_cat": p.get("Ownership_Category", ""),
        "lon": round(lon, 6) if lon else None,
        "lat": round(lat, 6) if lat else None,
        "overall_sqft": p.get("Overall_SQ_FT"),
        "built_status": p.get("Built_Status", ""),
        "zoning": p.get("ZONING", ""),
        "place_type": p.get("PLACE_TYPE", ""),
        "policy_area": p.get("POLICY_AREA", ""),
        "gis_acres": p.get("Existing_Acres"),
        "n_buildings": bldg.get("n_buildings"),
        "total_bldg_sqft": bldg.get("total_bldg_sqft"),
        "earliest_permit": bldg.get("earliest_permit"),
        "latest_permit": bldg.get("latest_permit"),
        "earliest_year": bldg.get("earliest_year"),
        "n_built": bldg.get("n_built"),
        "n_under_construction": bldg.get("n_under_construction"),
        "addresses": bldg.get("addresses", ""),
    }
    inventory.append(row)

# --- 7. Save CSV ---
out_file = OUT / "dc_master_inventory.csv"
fieldnames = list(inventory[0].keys())
with open(out_file, "w", newline="") as f:
    writer = csv.DictWriter(f, fieldnames=fieldnames)
    writer.writeheader()
    writer.writerows(inventory)

# --- 8. Summary ---
n_with_bldg = sum(1 for r in inventory if r["n_buildings"])
n_built = sum(1 for r in inventory if r["built_status"] == "BUILT")
n_uc = sum(1 for r in inventory if r["built_status"] == "UNDER CONSTRUCTION")
dates = [r["earliest_permit"] for r in inventory if r["earliest_permit"]]
total_bldg = sum(r["n_buildings"] or 0 for r in inventory)
total_sqft = sum(r["total_bldg_sqft"] or 0 for r in inventory)

zoning_counts = defaultdict(int)
for r in inventory:
    zoning_counts[r["zoning"]] += 1

print(f"\n{'='*50}")
print(f"DC MASTER INVENTORY")
print(f"{'='*50}")
print(f"Total parcels:           {len(inventory)}")
print(f"With building data:      {n_with_bldg}")
print(f"Without building data:   {len(inventory) - n_with_bldg}")
print(f"Built:                   {n_built}")
print(f"Under construction:      {n_uc}")
print(f"Total buildings:         {total_bldg}")
print(f"Total building sqft:     {total_sqft:,.0f}")
print(f"Permit date range:       {min(dates) if dates else 'N/A'} to {max(dates) if dates else 'N/A'}")
print(f"\nZoning breakdown:")
for z, c in sorted(zoning_counts.items(), key=lambda x: -x[1]):
    print(f"  {z}: {c}")
print(f"\nSaved to: {out_file}")
